### PR TITLE
`expval` ignores wire labels and uses position implicitly for tensor product ops

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
Consider the following circuit: 

```python 
dev = qml.device("default.qubit", wires=3)

@qml.qnode(dev)
def circ():
    qml.Hadamard(wires=0)
    qml.CNOT(wires=[0, 2])
    return qml.expval(qml.Identity(wires=0) @ qml.PauliZ(wires=1))      # -->  0.9999999999999998
    # return qml.expval(qml.Identity(wires=1) @ qml.PauliZ(wires=0))   # -->  0.9999999999999998
    # return qml.expval(qml.PauliZ(wires=0) @ qml.Identity(wires=1))   # -->  0.0
    # return qml.expval(qml.PauliZ(wires=1) @ qml.Identity(wires=0))   # -->  0.0

```

But the expected behaviour should be: 

```
... 
    return qml.expval(qml.Identity(wires=0) @ qml.PauliZ(wires=1))      # -->  0.9999999999999998
    # return qml.expval(qml.Identity(wires=1) @ qml.PauliZ(wires=0))   # -->  0.0
    # return qml.expval(qml.PauliZ(wires=0) @ qml.Identity(wires=1))   # -->  0.0
    # return qml.expval(qml.PauliZ(wires=1) @ qml.Identity(wires=0))   # -->  0.9999999999999998
```

It seems as though the wire labels are being ignored and instead the position of the operations in the tensor product are being used to determine how to compute the expectation value.

___Additionally:___ 
This PR also addresses the following bug which was discovered during the testing of this feature: 

```python
    dev = qml.device("default.qubit", wires=3)
    theta = 0.432
    phi = 0.123
    varphi = -0.543
 
    @qml.qnode(dev)
    def circuit(o):
        qfunc()
        return qml.expval(qml.apply(o))

    obs0 = qml.PauliZ(wires=label_map[2])

    obs1 = qml.Identity(wires=0) @ qml.Identity(wires=1) @ qml.PauliZ(wires=2)
    obs2 = qml.Identity(wires=1) @ qml.Identity(wires=0) @ qml.PauliZ(wires=2)

    obs3 = qml.Identity(wires=0) @ qml.PauliZ(wires=2) @ qml.Identity(wires=1)
    obs4 = qml.Identity(wires=1) @ qml.PauliZ(wires=2) @ qml.Identity(wires=0)

    obs5 = qml.PauliZ(wires=2) @ qml.Identity(wires=0) @ qml.Identity(wires=1)
    obs6 = qml.PauliZ(wires=2) @ qml.Identity(wires=1) @ qml.Identity(wires=0)

    obs_lst = [obs0, obs1, obs2, obs3, obs4, obs5, obs6]

    for i, o in enumerate(obs_lst):
        print(f"------ {i} ------- \n{circuit(o)}")
```

Which produces the following result: 
```
------ 0 -------
0.8561624160163042
------ 1 -------
0.8561624160163042
------ 2 -------
0.8561624160163042
------ 3 -------
0.8561624160163042
------ 4 -------
0.9081301906949608
------ 5 -------
0.9924450321351934
------ 6 -------
0.8561624160163042
```

In the above result we notice that the expectation values computed for observables 4, 5 do not match the others.

**Description of the Change:**
The `expval(self, observable, shot_range=None, bin_size=None)` function defined in `_qubit_device.py` uses the eigenvalues of the provided tensor product observable and the probability vector of the state to compute the expectation value when `shots=None`. 

The first bug arises because of lexicographic sorting of the eigvals  based on the wires the observables act on in the function `qml.Tensor.get_eigvals()`. This results in the eigvals of the two observables `PauliZ @ Identity` and `Identity @ PauliZ` being the same.  

This is addressed by removing the sorting in the `get_eigvals()` function. 

The second bug arises because of the way we permute the probability vector before computing the expectation value. The wire order of the observables is passed into the `device.probability()` function which then computes the marginal probability distribution over the observable wires and permutes the distribution according to the wire order of the observable wires. This permutation is incorrect and instead we should be using the inverse of this permutation. The reason such a bug was not caught sooner is because in most cases the tensor product observables usually contain 2 terms at most and/or there is only one qubit flipped out of position. In these common cases, the wire order permutation is its own inverse, and thus produces the correct expected value. But in situations with more then one qubit flip, it breaks. 

This is addressed by computing the permutation before hand, then permuting the wires before they are passed into the probability function. 
 
**Related GitHub Issues:**
This PR closes this [issue](https://github.com/PennyLaneAI/pennylane/issues/2263)

### CODE REVIEW

**Comment by Jaybsoni** [comment]
[sc-15486]

**Comment by Jaybsoni** [comment]
The reason some of the tests are failing is because of the following edge case: 

```
A = np.array( ... some hermitian 8x8 matrix ... )
H = qml.Hermitian(A, wires=[2,1,0])

def circ(): 
... some ops ...
    return qml.expval(H)   # now that we sort the wires for the probability vector but not for the Eigen_values, we get the "wrong" answer! 
```

The answer would be "correct" if we interpreted the matrix as already being transformed under the provided wire order.

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -65,6 +65,7 @@ jobs:
           pip install -r requirements-ci.txt
           python setup.py bdist_wheel
           pip install dist/PennyLane*.whl
+          pip install --upgrade --pre -i https://test.pypi.org/simple/ PennyLane-Lightning==0.22.0.dev16
```
Comment (Jaybsoni) [inline .github/workflows/tests.yml:68]:
```suggestion

```

File: pennylane/_qubit_device.py [hunk 2]
```diff
@@ -779,7 +779,21 @@ def expval(self, observable, shot_range=None, bin_size=None):
                     f"Cannot compute analytic expectations of {observable.name}."
                 ) from e
 
-            prob = self.probability(wires=observable.wires)
+            # the probability vector must be permuted to account for the permuted wire order of the observable
+            # one must be careful to realize that the observable wire order DOES NOT contain the same permutation
+            # of wire labels and instead, one must compute the permutation which maps the observable wire order back
+            # to the device wire order. This permutation, when applied to the probability vector will match the eigvals
+            # and ultimately compute the correct exp_val or variance
+
```
Comment (Jaybsoni) [inline pennylane/_qubit_device.py:787]:
```suggestion
```

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -25,10 +25,32 @@
...
+  
+  op1 = qml.Identity(wires=0) @ qml.Identity(wires=1) @ qml.PauliZ(wires=2)
+  op2 = qml.PauliZ(wires=2) @ qml.Identity(wires=0) @ qml.Identity(wires=1)
+  ```
+
+  ```
+  >>> print(circ(op1), circ(op2))
+  -0.8636111153905662 0.9928086358538663
+  ```
+
```
Comment (Jaybsoni) [inline doc/releases/changelog-dev.md:48]:
```suggestion
* Fixes a bug in which the `expval`/`var` of a `Tensor(Observable)` would depend on the order 
  in which the observable is defined: 
  ```python
  @qml.qnode(dev)
  def circ(op):
    qml.RX(0.12, wires=0)
    qml.RX(1.34, wires=1)
    qml.RX(3.67, wires=2)
    
    return qml.expval(op)
  
  op1 = qml.Identity(wires=0) @ qml.Identity(wires=1) @ qml.PauliZ(wires=2)
  op2 = qml.PauliZ(wires=2) @ qml.Identity(wires=0) @ qml.Identity(wires=1)
  ```

  ```
  >>> print(circ(op1), circ(op2))
  -0.8636111153905662 0.9928086358538663
  ```
  [(#2276)](https://github.com/PennyLaneAI/pennylane/pull/2276)
```

File: tests/test_measurements.py [hunk 4]
```diff
@@ -117,6 +147,38 @@ def circuit():
...
+        @qml.qnode(dev)
+        def circ(obs):
+            sub_routine(label_map=range(3))
+            return qml.expval(obs)
+
+        for obs, permuted_obs in zip(obs_lst, obs_permuted_lst):
+            assert np.allclose(circ(obs), circ(permuted_obs), atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("label_map", label_maps)
+    def test_wire_label_in_tensor_prod_observables(self, label_map, tol):
```
Comment (antalszava) [inline tests/test_measurements.py:162]: Is the difference between this test case and the previous one that the order of observables is fixed, but their wire labels change using parametrization?

**Review by antalszava — COMMENTED** [review]
Hey @Jaybsoni, amazing work! :fireworks: :1st_place_medal: I've left some comments, but no major blockers.

I think a bit more clarification could help (e.g., as an example in a docstring), for future maintenance.

Also: how about adding the test cases to the device test suite in a similar fashion as the 3-term test case was added in
https://github.com/PennyLaneAI/pennylane/pull/2292? That way these cases can be tested on any device that supports tensor observables.

File: pennylane/_device.py [hunk 5]
```diff
@@ -329,6 +329,30 @@ def define_wire_map(self, wires):
         wire_map = zip(wires, consecutive_wires)
         return OrderedDict(wire_map)
 
+    def get_ordered_subset(self, subset_wires):
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/_device.py:332]: I know there existing methods/functions that use `get`/`generate` in the codebase (\*cough\* `tape.get_parameters()` \*cough\*), _but_:

- I would like to ban new methods/functions that begin with `get` or `generate` unless it is absolutely necessary, it doesn't add much meaning to the function!

- In a similar vein, we should aim for 3 underscores maximum

File: doc/releases/changelog-dev.md [hunk 6]
```diff
@@ -42,10 +42,33 @@
...
+    return qml.expval(op)
+  
+  op1 = qml.Identity(wires=0) @ qml.Identity(wires=1) @ qml.PauliZ(wires=2)
+  op2 = qml.PauliZ(wires=2) @ qml.Identity(wires=0) @ qml.Identity(wires=1)
+  ```
+
+  ```
+  >>> print(circ(op1), circ(op2))
+  -0.8636111153905662 0.9928086358538663
+  ```
```
Comment (josh146) [inline doc/releases/changelog-dev.md:65]: Maybe it is better to show the behaviour _after_ the bugfix? That these now return the same results?

Comment (antalszava) [reply doc/releases/changelog-dev.md:65]: Agreed, was also wondering about that myself too!

Comment (Jaybsoni) [reply doc/releases/changelog-dev.md:65]: Ahh, sounds good! I guess I put the behaviour before because it was more illustrative of what the problem was. The behaviour afterwards is intuitive and expected, so its not really clear what the issue was. I will swap it over to the behaviour afterwards, and we can rely on the description above.

File: doc/releases/changelog-dev.md [hunk 7]
```diff
@@ -42,10 +42,33 @@
...
+    
+    return qml.expval(op)
+  
+  op1 = qml.Identity(wires=0) @ qml.Identity(wires=1) @ qml.PauliZ(wires=2)
+  op2 = qml.PauliZ(wires=2) @ qml.Identity(wires=0) @ qml.Identity(wires=1)
+  ```
+
+  ```
+  >>> print(circ(op1), circ(op2))
+  -0.8636111153905662 0.9928086358538663
```
Comment (Jaybsoni) [inline doc/releases/changelog-dev.md:64]:
```suggestion
  -0.8636111153905662 -0.8636111153905662
```

File: pennylane/_device.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/_device.py:332]: How about: `order_wires`?

File: pennylane/_qubit_device.py [hunk 8]
```diff
@@ -773,7 +773,20 @@ def expval(self, observable, shot_range=None, bin_size=None):
                     f"Cannot compute analytic expectations of {observable.name}."
                 ) from e
 
-            prob = self.probability(wires=observable.wires)
+            # the probability vector must be permuted to account for the permuted wire order of the observable
+            # one must be careful to realize that the observable wire order DOES NOT contain the same permutation
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/_qubit_device.py:777]: Sure, @antalszava should I add this to the comment? or is there a better place to highlight this?

[the remark this answers is not in the document]
Comment (antalszava) [reply pennylane/_qubit_device.py:777]: > Sure, @antalszava should I add this to the comment? or is there a better place to highlight this?

Personally, I'd create an auxiliary private method, in its docstring the toy example could reside nicely. Having said that, adding it to the comment here could also work nicely.

File: tests/test_measurements.py [hunk 9]
```diff
@@ -117,6 +147,38 @@ def circuit():
...
+        dev = qml.device("default.qubit", wires=3)
+        dev_custom_labels = qml.device("default.qubit", wires=[label_map[i] for i in range(3)])
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply tests/test_measurements.py:164]: You are correct haha ! I guess I must have been tired when I wrote that line :)

File: pennylane/_qubit_device.py [hunk 10]
```diff
@@ -773,7 +773,20 @@ def expval(self, observable, shot_range=None, bin_size=None):
...
+            # of wire labels and instead, one must compute the permutation which maps the observable wire order back
+            # to the device wire order. This permutation, when applied to the probability vector will match the eigvals
+            # and ultimately compute the correct exp_val or variance
+            ordered_obs_wire_lst = self.get_ordered_subset(observable.wires).tolist()
+            obs_wire_lst = observable.wires.tolist()
+
+            mapped_wires = [self.wire_map[label] for label in obs_wire_lst]
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/_qubit_device.py]: Added

File: tests/test_measurements.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (antalszava) [reply tests/test_measurements.py]: No worries :+1:

File: doc/releases/changelog-dev.md (hunk 6 above)
Comment (antalszava) [reply doc/releases/changelog-dev.md]: Minor here: apologies, from my side I wasn't clear: I think having both before and after could be nice. Not a blocking suggestion though.

File: pennylane/devices/tests/test_measurements.py [hunk 11]
```diff
@@ -411,6 +438,72 @@ def circuit():
...
+        "base_obs, permuted_obs",
+        list(zip(obs_lst, obs_permuted_lst)),
+    )
+    def test_wire_order_in_tensor_prod_observables(
+        self, device, base_obs, permuted_obs, tol, skip_if
+    ):
+        """Test that when given a tensor observable the expectation value is the same regardless of how the
+        observable is written, provided the wires it acts on are the same.
+
+        eg:
```
Comment (antalszava) [inline pennylane/devices/tests/test_measurements.py:451]:
```suggestion
        For example:

```

File: pennylane/devices/tests/test_measurements.py [hunk 12]
```diff
@@ -411,6 +438,72 @@ def circuit():
...
+            return qml.expval(ob)
+
+        assert np.allclose(circ(base_obs), circ(permuted_obs), atol=tol(dev.shots), rtol=0)
+
+    @pytest.mark.parametrize("label_map", label_maps)
+    def test_wire_label_in_tensor_prod_observables(self, device, label_map, tol, skip_if):
+        """Test that when given a tensor observable the expectation value is the same regardless of how the
+        wires are labelled, as long as they match the device order.
+
+        eg:
```
Comment (antalszava) [inline pennylane/devices/tests/test_measurements.py:477]:
```suggestion
        For example:

```

**Review by antalszava — APPROVED** [review]
Looks good :1st_place_medal: Amazing solution, goes a long way with understanding what's happening here.

Codefactor seems to cause some issues still and left come comments, but once those are addressed, this is good to be merged from my side.

File: pennylane/_device.py [hunk 13]
```diff
@@ -329,6 +329,30 @@ def define_wire_map(self, wires):
...
 
+    def order_wires(self, subset_wires):
+        """Given some subset of device wires return a Wires object with the same wires;
+        sorted according to the device wire map.
+
+        Args:
+            subset_wires (Wires): The subset of device wires (in any order).
+
+        Raise:
+            ValueError: Could not find some or all subset wires {subset_wires} in device wires {device_wires}.
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/_device.py:340]: I just included them to match the message raised by `ValueError`. But if it renders poorly then I agree we should remove it

File: pennylane/_qubit_device.py [hunk 14]
```diff
@@ -105,6 +105,60 @@ def _scatter(indices, array, new_dimensions):
...
+
+        .. code::
+            >>> print(0_1.wires)
+            <Wires = [0, 1, 2]>
+            >>> print(O_2.wires)
+            <Wires = [2, 0, 1]>
+
+        We might naively think that we must permute our state vector to match the wire order of our observable.
+        We must be careful and realize that the observable wire order DOES NOT contain the same permutation as the
+        observables themselves. If we directly compare :code:`O_1` and :code:`O_2`, we see that each sub-observable
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/_qubit_device.py]: Updated

File: pennylane/devices/tests/test_measurements.py [hunk 15]
```diff
@@ -411,6 +438,72 @@ def circuit():
...
+    )
+    def test_wire_order_in_tensor_prod_observables(
+        self, device, base_obs, permuted_obs, tol, skip_if
+    ):
+        """Test that when given a tensor observable the expectation value is the same regardless of how the
+        observable is written, provided the wires it acts on are the same.
+
+        eg:
+        ob1 = qml.PauliZ(wires=0) @ qml.PauliY(wires=1)
+        ob2 = qml.PauliY(wires=0) @ qml.PauliZ(wires=1)
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/devices/tests/test_measurements.py:454]: Ahh nice catch, the wire labels should be swapped 👍🏼

[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/devices/tests/test_measurements.py]: Added

File: pennylane/devices/tests/test_measurements.py [hunk 16]
```diff
@@ -411,6 +438,72 @@ def circuit():
...
 
+    @pytest.mark.parametrize(
+        "base_obs, permuted_obs",
+        list(zip(obs_lst, obs_permuted_lst)),
+    )
+    def test_wire_order_in_tensor_prod_observables(
+        self, device, base_obs, permuted_obs, tol, skip_if
+    ):
+        """Test that when given a tensor observable the expectation value is the same regardless of how the
+        observable is written, provided the wires it acts on are the same.
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/devices/tests/test_measurements.py]: Added

File: pennylane/devices/tests/test_measurements.py [hunk 17]
```diff
@@ -411,6 +438,72 @@ def circuit():
         expected = -(np.cos(varphi) * np.sin(phi) + np.sin(varphi) * np.cos(theta)) / np.sqrt(2)
         assert np.allclose(res, expected, atol=tol(dev.shots))
 
+    @pytest.mark.parametrize(
+        "base_obs, permuted_obs",
+        list(zip(obs_lst, obs_permuted_lst)),
+    )
+    def test_wire_order_in_tensor_prod_observables(
+        self, device, base_obs, permuted_obs, tol, skip_if
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/devices/tests/test_measurements.py:446]: It seems to be the preferred method by lint, should we overwrite it? I think its fine to leave it as is no ?
